src/client/websocket/WebSocketManager.js

```diff
@@ -37,52 +82,182 @@ class WebSocketManager extends EventEmitter {
-    if (!this.connection) {
-      this.debug('No connection to websocket');
-      return;
+  connect(gateway = this.gateway) {
+    this.gateway = gateway;
+
+    if (typeof this.client.options.shards === 'number') {
+      this.debug('Spawning 1 shard');
+      this.spawn(this.client.options.shards);
+    } else if (Array.isArray(this.client.options.shards)) {
+      this.debug(`Spawning ${this.client.options.shards.length} shards`);
+      for (let i = 0; i < this.client.options.shards.length; i++) {
+        this.spawn(this.client.options.shards[i]);
+      }
+    } else {
+      this.debug(`Spawning ${this.client.options.shardCount} shards`);
+      for (let i = 0; i < this.client.options.shardCount; i++) {
+        this.spawn(i);
+      }
     }
-    this.connection.send(packet);
   }
 
   /**
-   * Connects the client to a gateway.
-   * @param {string} gateway The gateway to connect to
+   * Processes a packet and queues it if this WebSocketManager is not ready.
+   * @param {Object} packet The packet to be handled
+   * @param {WebSocketShard} shard The shard that will handle this packet
    * @returns {boolean}
    */
-  connect(gateway) {
-    if (!this.connection) {
-      this.connection = new WebSocketConnection(this, gateway);
-      return true;
-    }
-    switch (this.connection.status) {
-      case Status.IDLE:
-      case Status.DISCONNECTED:
-        this.connection.connect(gateway, 5500);
-        return true;
-      default:
-        this.debug(`Couldn't connect to ${gateway} as the websocket is at state ${this.connection.status}`);
+  handlePacket(packet, shard) {
+    if (packet && this.status !== Status.READY) {
+      if (!BeforeReadyWhitelist.includes(packet.t)) {
+        this.packetQueue.push({ packet, shardID: shard.id });
         return false;
+      }
+    }
+
+    if (this.packetQueue.length) {
+      const item = this.packetQueue.shift();
+      this.client.setImmediate(() => {
+        this.handlePacket(item.packet, this.shards[item.shardID]);
+      });
+    }
+
+    if (packet && PacketHandlers[packet.t]) {
+      PacketHandlers[packet.t](this.client, packet, shard);
+    }
+
+    return false;
+  }
+
+  /**
+   * Checks whether the client is ready to be marked as ready.
+   * @returns {void}
+   */
+  checkReady() {
+    if (!(this.shards.filter(s => s).length === this.client.options.actualShardCount) ||
+      !this.shards.filter(s => s).every(s => s.status === Status.READY)) {
+      return false;
+    }
+
+    let unavailableGuilds = 0;
+    for (const guild of this.client.guilds.values()) {
+      if (!guild.available) unavailableGuilds++;
+    }
+    if (unavailableGuilds === 0) {
+      this.status = Status.NEARLY;
+      if (!this.client.options.fetchAllMembers) return this.triggerReady();
+      // Fetch all members before marking self as ready
+      const promises = this.client.guilds.map(g => g.members.fetch());
+      Promise.all(promises)
+        .then(() => this.triggerReady())
```

Comment:
Would I not be missing out on errors to catch if I swapped the order? In this way, it will catch all o them, including ones from the success callback. Or is that the intention here?